plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Types.hs

```diff
@@ -62,22 +79,61 @@ hasTests :: Section -> Bool
 hasTests = not . null . sectionTests
 
 hasPropertyTest :: Section -> Bool
-hasPropertyTest = any (isProperty . unLoc) . sectionTests
+hasPropertyTest = any isProperty . sectionTests
 
 -- |Split setup and normal sections
 splitSections :: [Section] -> ([Section], [Section])
 splitSections = partition ((== "setup") . sectionName)
 
 data Test
-    = Example {testLines :: NonEmpty Txt, testOutput :: [Txt]}
-    | Property {testline :: Txt, testOutput :: [Txt]}
+    = Example {testLines :: NonEmpty Txt, testOutput :: [Txt], testRange :: Range}
+    | Property {testline :: Txt, testOutput :: [Txt], testRange :: Range}
     deriving (Eq, Show, Generic, FromJSON, ToJSON, NFData)
 
+data Comments = Comments
+    { lineComments :: Map Range RawLineComment
+    , blockComments :: Map Range RawBlockComment
+    }
+    deriving (Show, Eq, Ord, Generic)
```

Comment:
Out of curiosity, does this work well with nested comments?

Reply:
Ah, I've never thought of such a situation.

It seems that if the eval line isn't placed right before the end of the inner ending brace, it seems work:

![nested-space](https://user-images.githubusercontent.com/21860/106018199-9d321280-6104-11eb-9c0f-883ecc598715.gif)

But if it is placed right before the inner closing brace, it will break the output (as it is considered as a result line literally saying `-}` and get replaced with evaluation result):

![nested-no-space](https://user-images.githubusercontent.com/21860/106018242-a7541100-6104-11eb-8112-25f525e5dcdb.gif)

It seems that old parser can handle both cases correctly, so it might be desirable to support such a case in this PR.
I think we don't have to maintain nesting level but suffices to treat line solely consisting of `-}` or `{-` and white spaces as a delimiting empty line in block comment contexts.

Indeed, there can be (somewhat pathological) cases that `-}` can appear as an evaluation result, e.g:

```hs
data EndingBrace = EndingBrace
instance Show EndingBrace where
  showsPrec _ = const $ showString "-}"

{-
>>> EndingBrace
-}
```

But this already breaks block comment eval line regardless of special treatment of `-}`-line, so mentioned special treatment of non-terminating `-}` and `{-` would suffice, I think.

Reply:
Thank you for the detailed investigation. How does the `doctest` tool work on `EndingBrace` case? It would be more safe to say "stable" if we can follow their practice, since their practice has a much longer history than the Eval plugin.

Reply:
> It would be more safe to say "stable" if we can follow their practice, since their practice has a much longer history than the Eval plugin.

Agreed. But we have to mind the major difference between Eval and doctest that the former has the functionality to evaluate line and insert to the source code, `doctest` does evaluation but just compares the result to existing contents following the test-line.

As for nested comments, `doctest` seems to ignore any eval inside nested block comments (of nesting level >= 2).
As for EndingBrace, it just tries to compare the newly-computed value with the existing one.
And doctest seems to treat escaped inputs such as `-\}` as-is, so it seems there is no way to check the validity of outputs containing `-}`.

That said, `EndingBrace` is just a corner-case and it seems such a case rarely comes up in practice.
If one has to use Eval Plugin to inspect output containing `-}`, they can use one-line comments instead anyway (plus some attention to configure code formatters not to convert them into multiline comments). So I think just ignoring nested `-}` makes sense in this case.

Reply:
> That said, EndingBrace is just a corner-case and it seems such a case rarely comes up in practice.

It indeed looks it's the case. Thank you for fixing this parser hazard!